src/bors/mergeable_queue.rs

```diff
@@ -0,0 +1,316 @@
+/// Exponential backoff delay multiplier.
+const BACKOFF_MULTIPLIER: f64 = 2.0;
+/// Max number of mergeable check retries before giving up.
+const MAX_RETRIES: u32 = 5;
+
+#[derive(Debug, Clone)]
+pub struct QueuedPullRequest {
+    pub pr_number: PullRequestNumber,
+    pub repo: GithubRepoName,
+}
+
+impl std::fmt::Display for QueuedPullRequest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}/#{}", self.repo, self.pr_number)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct MergeableQueueItem {
+    pub pull_request: QueuedPullRequest,
+    pub attempt: u32,
+    pub immediate: bool,
+}
+
+#[derive(Debug, Clone)]
+struct Item {
+    /// When to process item (None = immediate).
+    /// Reversed to create min-heap for expirations.
+    expiration: Reverse<Option<Instant>>,
+    inner: MergeableQueueItem,
+}
+
+impl PartialEq for Item {
+    fn eq(&self, other: &Self) -> bool {
+        self.expiration == other.expiration
+    }
+}
+
+impl Eq for Item {}
+
+impl PartialOrd for Item {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Item {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.expiration.cmp(&other.expiration)
+    }
+}
+
+struct SharedInner {
+    queue: Mutex<BinaryHeap<Item>>,
+    notify: Notify,
+    closed: AtomicBool,
+}
+
+#[derive(Clone)]
+pub struct MergeableQueueSender {
+    inner: Arc<SharedInner>,
+}
+
+#[derive(Clone)]
+pub struct MergeableQueueReceiver {
+    inner: Arc<SharedInner>,
+}
+
+pub fn create_mergeable_queue() -> (MergeableQueueSender, MergeableQueueReceiver) {
+    let shared = Arc::new(SharedInner {
+        queue: Mutex::new(BinaryHeap::new()),
+        notify: Notify::new(),
+        closed: AtomicBool::new(false),
+    });
+
+    (
+        MergeableQueueSender {
+            inner: shared.clone(),
+        },
+        MergeableQueueReceiver { inner: shared },
+    )
+}
+
+impl MergeableQueueSender {
+    pub fn close(&self) {
+        self.inner.closed.store(true, Ordering::Relaxed);
+        self.inner.notify.notify_waiters();
+    }
+
+    pub fn enqueue(&self, repo: GithubRepoName, pr_number: PullRequestNumber) {
+        let expiration = Some(Instant::now() + BASE_DELAY);
+
+        self.insert_item(
+            MergeableQueueItem {
+                pull_request: QueuedPullRequest { pr_number, repo },
+                attempt: 1,
+                immediate: false,
+            },
+            expiration,
+        );
+    }
+
+    #[allow(dead_code)]
+    pub fn enqueue_now(&self, pr_number: PullRequestNumber, repo: GithubRepoName) {
+        self.insert_item(
+            MergeableQueueItem {
+                pull_request: QueuedPullRequest { pr_number, repo },
+                attempt: 1,
+                immediate: true,
+            },
+            None,
+        );
+    }
+
+    pub fn enqueue_retry(&self, queue_item: MergeableQueueItem) {
```

Comment:
```suggestion
    fn enqueue_retry(&self, queue_item: MergeableQueueItem) {
```
This should be private (at least until the processing logic is still in this file).